pkg/sfu/connectionquality/connectionstats.go

```diff
@@ -293,32 +147,58 @@ func (cs *ConnectionStats) updateStatsWorker() {
 	defer tk.Stop()
 
 	for {
-		select {
-		case <-cs.done:
+		<-tk.C
+
+		if cs.isClosed.Load() {
 			return
+		}
 
-		case <-tk.C:
-			stat := cs.getStat()
-			if stat == nil {
-				continue
-			}
+		stat := cs.getStat(time.Now())
+		if stat == nil {
+			continue
+		}
 
-			if cs.onStatsUpdate != nil {
-				cs.onStatsUpdate(cs, stat)
-			}
+		if cs.onStatsUpdate != nil {
+			cs.onStatsUpdate(cs, stat)
 		}
 	}
 }
 
 // -----------------------------------------------------------------------
 
-func getCodecNameFromMime(mime string) string {
-	codecName := ""
-	codecParsed := strings.Split(strings.ToLower(mime), "/")
-	if len(codecParsed) > 1 {
-		codecName = codecParsed[1]
+// how much weight to give to packet loss rate when calculating score.
+// It is codec dependent.
+// For audio:
+//   o Opus without FEC or RED suffers the most through packet loss, hence has the highest weight
+//   o RED with two packet redundancy can absorb two out of every three packets lost, so packet loss is not as detrimental and therefore lower weight
+//
+// For video:
+//   o No in-built codec repair available, hence same for all codecs
+func getPacketLossWeight(mimeType string, isFecEnabled bool) float64 {
```

Comment:
how were these constants picked?

Reply:
Empirical. There is the e-model, an ITU-T standard, but not tuned for modern codecs. rtcscore was an attempt at a model, but that is a bit complex and also does not consider packet loss for video codecs. So, this is a result of filtering a bunch of those inputs/ signals into something simple.

Reply:
yeah i assumed as much i was just asking out of curiosity. red is like 2 or 3x redundant broadcasts? and fec is some sort of error correcting code. do the qualities represent some probability of having undecidable segments of some length within a sample?

Reply:
- rtcscore: https://github.com/ggarber/rtcscore
- Shishir ported that to Go: https://github.com/livekit/rtcscore-go

Reply:
> yeah i assumed as much i was just asking out of curiosity. red is like 2 or 3x redundant broadcasts? and fec is some sort of error correcting code. do the qualities represent some probability of having undecodable segments of some length within a sample?

Yes, these are representative. Ideally, we would get data from the edge about the following
- how many packets/samples had to be concealed.
- how many were recovered using FEC.
But, those are harder to do. For example, an audio track could get forwarded to multiple subscribers. Each subscriber could behave differently. To get purely upstream quality, SFU would have to try to decode (including a jitter buffer) to really figure out what is the quality of the upstream track.

RED used to be 3x (every packet includes 2 previous packets), but I thought I saw 2 packets recently. Not sure if something has changed. If it is 3, we can lose two packets out of every three and still be able to decode 100%. But, losses are bursty and quality is not great beyond 10% empirically (from user reports). We can measure loss characteristic (isolated vs bursty) and apply a better weight based on pattern in the analysis window. But, that is more compute. We do measure that for the whole stream, but been avoiding doing that in analysis windows as it is per packet processing.

Opus has in-built FEC. It adds FEC bits based on loss. Based on loss reported in RTCP Receiver Report, Opus will add FEC bits if negotiated. Again, it does not do great against bursty losses, but does quite well when having isolated losses (https://opus-codec.org/examples/).

So, this combination of numbers is a filtering of all the factors and is a balance between available data/available knowledge on this matter (which I am probably privy to only 1% or less)/implementation considerations.